allennlp/modules/transformer/attention_module.py

```diff
@@ -0,0 +1,608 @@
+        )
+        values = self.relative_attention_bias(
+            relative_position_bucket
+        )  # shape (query_length, key_length, num_heads)
+        values = values.permute([2, 0, 1]).unsqueeze(
+            0
+        )  # shape (1, num_heads, query_length, key_length)
+        return values
+
+
+class T5Attention(AttentionModule):
+
+    _pretrained_relevant_module = ["encoder.block.0.layer.0.SelfAttention"]
+    _pretrained_mapping = {
+        "q": "query",
+        "k": "key",
+        "v": "value",
+        "o": "output",
+    }
+
+    def __init__(
+        self,
+        is_decoder: bool = False,
+        hidden_size: int = 512,
+        key_value_proj_dim: int = 64,
+        num_heads: int = 8,
+        has_relative_attention_bias: bool = False,
+        relative_attention_num_buckets: int = 32,
+        dropout: float = 0.1,
+        normalize: bool = True,
+        is_cross_attention: bool = False,
+    ):
+
+        super().__init__(
+            hidden_size=hidden_size,
+            attention_head_size=key_value_proj_dim,
+            num_attention_heads=num_heads,
+            output_linear=True,
+            scoring_func="scaled_dot_product",
+            dropout=dropout,
+            bias=False,
+            normalize_weights=normalize,
+            is_decoder=is_decoder,
+            is_cross_attention=is_cross_attention,
+            has_relative_attention_bias=has_relative_attention_bias,
+            relative_attention_num_buckets=relative_attention_num_buckets,
+        )
+
+        self.attn = Attention.by_name(self.scoring_func)(scaling_factor=1, normalize=False)
+
+    def forward(  # type: ignore
+        self,
+        hidden_states: torch.Tensor,
+        mask: Optional[torch.BoolTensor] = None,
+        key_value_states: Optional[FloatT] = None,
+        position_bias: Optional[FloatT] = None,
+        past_key_value: Optional[
+            Tuple[FloatT, FloatT]
+        ] = None,  # this is used when taking decoding steps.
+        layer_head_mask: Optional[BoolT] = None,
+        query_length: Optional[int] = None,  # only relevant in cross-attention.
+        use_cache: bool = False,
+        output_attentions: bool = False,
+    ) -> AttentionOutput:
+        """
+        Self-attention (if key_value_states is None) or attention over source sentence (provided by
+        key_value_states).
+        """
+        if past_key_value:
+            past_key_states = past_key_value[0]
+            past_value_states = past_key_value[1]
+        else:
+            past_key_states = None
+            past_value_states = None
+
+        outputs = super().forward(
+            query_states=hidden_states,
+            past_key_states=past_key_states,
+            past_value_states=past_value_states,
+            attention_mask=mask,
+            source_states=key_value_states,
+            source_attention_mask=None,  # TODO: is this a bug in current T5 code?
```

Comment:
Maybe not for this PR, but we could chase this down? Talk to the original T5 people, or whoever implemented it for HF?